src/ASM/LR/ASMu3Dmxrecovery.C

```diff
@@ -0,0 +1,178 @@
+// $Id$
+//==============================================================================
+//!
+//! \file ASMu3Dmxrecovery.C
+//!
+//! \date Nov 9 2017
+//!
+//! \author Arne Morten Kvarving / SINTEF
+//!
+//! \brief Recovery techniques for unstructured mixed LR B-splines.
+//!
+//==============================================================================
+
+#include "LRSpline/LRSplineVolume.h"
+#include "LRSpline/Element.h"
+
+#include "ASMu3Dmx.h"
+#include "IntegrandBase.h"
+#include "CoordinateMapping.h"
+#include "GaussQuadrature.h"
+#include "SparseMatrix.h"
+#include "SplineUtils.h"
+#include "Profiler.h"
+#include "matrix.h"
+#include <numeric>
+
+
+/*!
+  \brief Expands a tensor parametrization point to an unstructured one.
+*/
+
+static void expandTensorGrid (const RealArray* in, RealArray* out)
+{
+  out[0].resize(in[0].size()*in[1].size()*in[2].size());
+  out[1].resize(in[0].size()*in[1].size()*in[2].size());
+  out[2].resize(in[0].size()*in[1].size()*in[2].size());
+
+  size_t i, j, k, ip = 0;
+  for (k = 0; k < in[2].size(); k++)
+    for (j = 0; j < in[1].size(); j++)
+      for (i = 0; i < in[0].size(); i++, ip++) {
+        out[0][ip] = in[0][i];
+        out[1][ip] = in[1][j];
+        out[2][ip] = in[2][k];
+      }
+}
+
+
+bool ASMu3Dmx::assembleL2matrices (SparseMatrix& A, StdVector& B,
+                                   const IntegrandBase& integrand,
+                                   bool continuous) const
+{
+  const int p1 = projBasis->order(0);
+  const int p2 = projBasis->order(1);
+  const int p3 = projBasis->order(2);
+
+  // Get Gaussian quadrature points
+  const int ng1 = continuous ? nGauss : p1 - 1;
+  const int ng2 = continuous ? nGauss : p2 - 1;
+  const int ng3 = continuous ? nGauss : p3 - 1;
+  const double* xg = GaussQuadrature::getCoord(ng1);
+  const double* yg = GaussQuadrature::getCoord(ng2);
+  const double* zg = GaussQuadrature::getCoord(ng3);
+  const double* wg = continuous ? GaussQuadrature::getWeight(nGauss) : nullptr;
+  if (!xg || !yg || !zg) return false;
+  if (continuous && !wg) return false;
+
+  size_t nnod = this->getNoProjectionNodes();
+  double dV = 0.0;
+  Vectors phi(2);
+  Matrices dNdu(2);
+  Matrix sField, Xnod, Jac;
+  std::vector<Go::BasisDerivs> spl1(2);
+  std::vector<Go::BasisPts> spl0(2);
+
+
+  // === Assembly loop over all elements in the patch ==========================
+  LR::LRSplineVolume* geoVol;
+  if (m_basis[geoBasis-1]->nBasisFunctions() == projBasis->nBasisFunctions())
+    geoVol = m_basis[geoBasis-1].get();
+  else
+    geoVol = projBasis.get();
+
+  for (auto el1 : geoVol->getAllElements())
```

Comment:
see previous comment on the use of `auto`

Reply:
fixed.